src/main/java/ru/curs/xylophone/MergeRegionContainer.java

```diff
@@ -0,0 +1,230 @@
+                    String.format("Cannot merge cell with address %s. It is out of range for up merge",
+                            address.getAddress()));
+        }
+        CellRangeAddress res = new CellRangeAddress(address.getRow() - 2,
+                address.getRow() - 1, address.getCol() - 1, address.getCol() - 1);
+        Optional<MutableCellRangeAddress> intersectedRegion = findIntersectedRegion(res);
+
+        if (intersectedRegion.isPresent()) {
+            mergeIntersectedRegions(intersectedRegion.get(), res);
+        } else {
+            mergedRegions.add(new MutableCellRangeAddress(res));
+        }
+    }
+
+    /**
+     * Merge with left cell.
+     * @param address address of current cell.
+     */
+    public void mergeLeft(CellAddress address) {
+        if (address.getCol() == 1) {
+            throw new IllegalArgumentException(
+                    String.format("Cannot merge cell with address %s. It is out of range for left merge",
+                            address.getAddress()));
+        }
+        CellRangeAddress res = new CellRangeAddress(address.getRow() - 1,
+                address.getRow() - 1, address.getCol() - 2, address.getCol() - 1);
+        Optional<MutableCellRangeAddress> intersectedRegion = findIntersectedRegion(res);
+
+        if (intersectedRegion.isPresent()) {
+            mergeIntersectedRegions(intersectedRegion.get(), res);
+        } else {
+            mergedRegions.add(new MutableCellRangeAddress(res));
+        }
+    }
+
+    /**
+     * Add merged region.
+     * Static merge or iteration merge.
+     * @param res merge region
+     */
+    public void addMergedRegion(CellRangeAddress res) {
+        Optional<MutableCellRangeAddress> intersectedRegion = findIntersectedRegion(res);
+        if (intersectedRegion.isPresent()) {
+            mergeIntersectedRegions(intersectedRegion.get(), res);
+        } else {
+            mergedRegions.add(new MutableCellRangeAddress(res));
+        }
+    }
+
+    /**
+     * Apply all merged regions to sheet.
+     * @param sheet sheet of docs
+     */
+    public void apply(Sheet sheet) {
+        if (sheet == null) {
+            return;
+        }
+        mergedRegions.stream().map(MutableCellRangeAddress::toCellRangeAddress).forEach(sheet::addMergedRegion);
+        clear();
+    }
+
+    /**
+     * Find main region with current cell.
+     * For defining value in region.
+     * @param res range of current merge region.
+     * @return Current merge region if intersects is false; otherwise - intersected existing region.
+     */
+    public CellRangeAddress findIntersectedRange(CellRangeAddress res) {
+        if (res.getFirstColumn() < 0 || res.getFirstRow() < 0) {
+            throw new IllegalArgumentException("Cannot merge first row with upper cell or first column with left cell");
+        }
+        Optional<MutableCellRangeAddress> intersected = findIntersectedRegion(res);
+        if (intersected.isPresent()) {
```

Comment:
`isPresent()` usage is generally a sign of something is wrong ))

should be replaced with 

```
return intersected.map(MutableCellRangeAddress::toCellRangeAddress).orElse(res)
```